File: java/binary/src/main/java/Binary.java
import java.util.stream.IntStream;

public class Binary {

    private static final String VALID_BINARY = "[01]+";

    private final int decimal;

    public Binary(final String binary) {
        if (binary.matches(VALID_BINARY)) {
            char[] binaryDigits = new StringBuilder(binary).reverse().toString().toCharArray();

            decimal = IntStream.range(0, binaryDigits.length)
                    .mapToObj(i -> Character.getNumericValue(binaryDigits[i]) * Math.pow(2, i))
                    .mapToInt(Double::intValue)
                    .sum();
        } else {
            decimal = 0;
        }
    }

    public int getDecimal() {
        return decimal;
    }
}

File: ecmascript/binary/binary.js
export default class Binary {
  constructor(binary) {
    this.binary = binary;
  }

  toDecimal() {
    return BINARY.test(this.binary)
      ? [...this.binary].reverse().reduce((total, b, index) => total + b * Math.pow(2, index), 0)
      : 0;
  }
}

const BINARY = /^[01]+$/;
